src/ansys/fluent/core/session.py

```diff
@@ -343,8 +306,93 @@ def download(self, file_name: str, local_file_path: Optional[str] = None):
         FileNotFoundError
             If the remote file does not exist.
         """
-        if self.file_service:
-            if self.file_service.file_exist(file_name):
-                self.file_service.download_file(file_name, local_file_path)
+        file_service = self._file_service()
+        if file_service:
+            if file_service.file_exist(file_name):
+                file_service.download_file(file_name, local_file_path)
             else:
                 raise FileNotFoundError("Remote file does not exist.")
+
+    def _wait_for_file(self, file_name: str):
+        """Wait for file to get ready for upload or download.
+
+        Parameters
+        ----------
+        file_name : str
+            Case file name
+        Raises
+        ------
+        FileNotFoundError
+            If a case file does not exist.
+        """
+        file_service = self._file_service()
+        start_time = time.time()
+        max_wait_time = 100
+        while (time.time() - start_time) < max_wait_time:
+            if file_service.file_exist(os.path.basename(file_name)):
+                break
+            max_wait_time -= 1
+            time.sleep(3)
+        else:
+            raise FileNotFoundError(f"{file_name} does not exist.")
+
+    def _pypim_upload_helper(self, file_name, is_meshing, api):
+        """Uploads a case file if not available on the server.
+
+        Parameters
+        ----------
+        file_name : str
+            Case file name
+        is_meshing: bool
+            True if mode is meshing, otherwise False
+        api: Session object property
+            either session.tui or session.file
+        Raises
+        ------
+        FileNotFoundError
+            If a case file does not exist.
+        """
+        file_service = self._file_service()
+        if os.path.isfile(file_name):
+            if not file_service.file_exist(os.path.basename(file_name)):
+                self.upload(os.path.basename(file_name))
```

Comment:
Hi @hpohekar I just want to double-check this logic.

We check if file with given path exists locally 
Then (after the remote check), we upload by base name only, i.e. stripping off the path information. How can `upload()` know where the file resides locally?

Am I understanding correctly?

Reply:
@seanpearsonuk Yes, got it. In the Ansys Lab I have `elbow.cas.h5` in both current directory and in `read/elbow.cas.h5` directory. So every time it is taking `elbow.cas.h5` from current directory after stripping off the path information.

It should be `self.upload(file_name)` and `file_name` can be absolute or relative path.

Thank you very much, for pointing this out.

I will update this change.

Reply:
Done and verified. Thank you.